assignments/introductory-assignments.md

```diff
@@ -488,14 +582,767 @@ A digital circuit computer with 2GB of memory can distinguish 2^{17 179 869 184}
+
+How do we get to such numbers? 
+
+When write 4 GB in scientific notation, we get: 
+
+2^{2} \* 2^{30} \* 2{3} = 2{35}
+
+Where: 
+
+- Quantity of 4 = 2^{2}, 
+- 2^{30} stands for Giga and 
+- 2^{3} stands for 8 when we convert from bytes to bits.
+
+Now we know that 4 GB of memory is 2^{35} bit, but how is that large memory split up into smaller pieces?
+
+Most registers of a CPU have the same size, that is, accommodate the same number of bits. Usually, data goes back and forth between memory and registers in chunks of that size called machine words or just words.
+
+[Word](https://en.wikipedia.org/wiki/Word_(computer_architecture)): The natural unit of data used by a particular processor design. A word is a fixed-sized piece of data handled as a unit by the instruction set or the processor's hardware. The number of bits in a word (the word size, word width, or word length) is an essential characteristic of any specific processor design or computer architecture.
+
+Usually, a word has 32 bit, as we can see in `selfie.c`. That means a double word has 64 bit, but for simplification, when we talk about a word in Selfie, we mean 64 bit.
+
+The processor (CPU) that the mipster emulator in Selfie implements has a size of 64 bits. Virtually everything on that machine happens at the level of words. Loading data from memory and storing it back, arithmetic and logical operations among registers, and even fetching code from memory for execution. The reason is again performance. Involving 64 bits in parallel in all operations is faster than working with bits individually. However, there are two more reasons why we use a 64-bit machine. The first reason is that the size of an integer in C\* (a subset of C) is also 64 bits. That means that a single mipster register can hold precisely one C\* integer value.
+
+[Emulator](https://en.wikipedia.org/wiki/Emulator): Software that enables one computer system (called the host) to behave like another computer system (called the guest).
+
+How many different integer values can 64 bits represent? Well, 2^{64} values, but what are they? 
+That depends on how we interpret them. In C\* integers are interpreted as *unsigned*, an integer value is either zero or a positive number.
+
+> An unsigned integer in C\* can in total represent integer values i from 0 to 18446744073709551615 = 2^{64}-1. 
+> In `selfie.c`, that value is called `UINT64_MAX` which is the largest unsigned 64-bit integer value.
+
+However, we may also interpret 64-bit integers as signed in *two's complement* with the MSB encoding the sign and the remaining 63 LSBs encoding the value. The number of different integer values that can be represented is nevertheless the same.
+
+> A signed 64-bit integer can in total represent signed integer values i from -9223372036854775808 to 9223372036854775807 since -2^{64-1} to 2^{64-1}-1. 
```

Comment:
```suggestion
> A signed 64-bit integer can in total represent values from -9223372036854775808 to 9223372036854775807 since -2^{64-1} to 2^{64-1}-1. 
```